rust/automerge/src/transaction/inner.rs

```diff
@@ -447,30 +521,142 @@ impl TransactionInner {
+        obj: ObjId,
+        mut index: usize,
+        mut del: usize,
+        vals: Vec<ScalarValue>,
+        splice_type: SpliceType<'_>,
+    ) -> Result<(), AutomergeError> {
+        let ex_obj = doc.ops.id_to_exid(obj.0);
+        let encoding = splice_type.encoding();
+        // delete `del` items - performing the query for each one
+        let mut deleted = 0;
+        while deleted < del {
+            // TODO: could do this with a single custom query
+            let query = doc.ops.search(&obj, query::Nth::new(index, encoding));
+
+            // if we delete in the middle of a multi-character
+            // move cursor back to the beginning and expand the del width
+            let adjusted_index = query.index();
+            if adjusted_index < index {
+                del += index - adjusted_index;
+                index = adjusted_index;
+            }
+
+            let step = if let Some(op) = query.ops.last() {
+                op.width(encoding)
             } else {
-                self.local_op::<Obs>(doc, None, obj, pos.into(), OpType::Delete)?;
+                break;
+            };
+
+            let op = self.next_delete(query.key()?, query.pred(&doc.ops));
+
+            doc.ops.add_succ(&obj, &query.ops_pos, &op);
+
+            self.operations.push((obj, op));
+
+            deleted += step;
+        }
+
+        if deleted > 0 {
+            if let Some(obs) = op_observer.as_mut() {
+                obs.delete_seq(doc, ex_obj.clone(), index, deleted);
             }
         }
-        for v in vals {
-            // As above this unwrap and rewrap of the option is necessary to appeas the borrow checker :(
+
+        // do the insert query for the first item and then
+        // insert the remaining ops one after the other
+        if !vals.is_empty() {
+            let query = doc.ops.search(&obj, query::InsertNth::new(index, encoding));
+            let mut pos = query.pos();
+            let mut key = query.key()?;
+            let mut cursor = index;
+            let mut width = 0;
+
+            for v in &vals {
+                let op = self.next_insert(key, v.clone());
+
+                doc.ops.insert(pos, &obj, op.clone());
+
+                width = op.width(encoding);
+                cursor += width;
+                pos += 1;
+                key = op.id.into();
+
+                self.operations.push((obj, op));
+            }
+
+            doc.ops.hint(&obj, cursor - width, pos - 1);
+
+            // handle the observer
             if let Some(obs) = op_observer.as_mut() {
-                self.do_insert(doc, Some(*obs), obj, pos, v.clone().into())?;
-            } else {
-                self.do_insert::<Obs>(doc, None, obj, pos, v.clone().into())?;
+                match splice_type {
+                    SpliceType::List => {
+                        let start = self.operations.len() - vals.len();
+                        for (offset, v) in vals.iter().enumerate() {
```

Comment:
Can we express this loop as `for (op, value) in self.operations.skip(start).zip(vals.iter())` or similar?

Reply:
I think what it needs is an `observer.insert()` that takes many values.  Currently text can be spliced but no lists.